spec/src/main/asciidoc/repository.asciidoc

```diff
@@ -112,6 +112,246 @@ Subsequently, an implementation of Jakarta Data, specifically tailored to the ch
+
+The subsections below specify the rules that an abstract method declaration must observe so that the Jakarta Data implementation is able to provide an implementation of the abstract method.
+
+- If every abstract method of a repository complies with the rules specified below, then the Jakarta Data implementation must provide an implementation of the repository.
+- Otherwise, if a repository declares an abstract method which does not comply with the rules specified below, or makes use of functionality which is not supported by the Jakarta Data implementation, then an error might be produced by the Jakarta Data implementation at build time or at runtime.
+
+The portability of a given repository interface between Jakarta Data implementations depends on the portability of the entity types it uses.
+If an entity class is not portable between given implementations, then any repository which uses the entity class is also unportable between those implementations.
+
+[NOTE]
+Additional portability guarantees may be provided by specifications which extend this specification, specializing to a given class of datastore.
+
+==== Lifecycle methods
+
+A _lifecycle method_ is an abstract method annotated with a _lifecycle annotation_.
+Lifecycle methods allow the program to make changes to persistent data in the data store.
+
+A lifecycle method must:
+
+- have a single parameter annotated with a lifecycle annotation, whose type is a concrete entity class `E`, and
+- be declared `void`, or return exactly the same entity type `E`.
+
+That is, its signature must be:
+
+[source,java]
+----
+void lifecycle(@Lifecycle Entity e);
+----
+
+or:
+
+[source,java]
+----
+Entity lifecycle(@Lifecycle Entity e);
+----
+
+where `lifecycle` is the arbitrary name of the method, `Entity` is a concrete entity class, and `Lifecycle` is a lifecycle annotation.
+
+This specification defines three built-in lifecycle annotations: `@Insert`, `@Update`, and `@Delete`.
+
+For example:
+
+[source,java]
+----
+void insertBook(@Insert Book book);
+----
+
+Lifecycle methods are not guaranteed to be portable between all providers.
+
+[NOTE]
+Any Jakarta Data implementation is strongly encouraged to support lifecycle methods declared using the built-in lifecycle annotations, to enable portability of user code between Jakarta Data providers.
+However, since these lifecycle operations are not universal across all possible data stores, support for `@Insert`, `@Update`, and `@Delete` is not required by this specification.
+
+There is no special programming model for lifecycle annotations.
+The Jakarta Data implementation automatically recognizes the lifecycle annotations it supports.
+
+[NOTE]
+====
+A Jakarta Data provider might extend this specification to define additional lifecycle annotations, or to support lifecycle methods with signatures other than the usual signatures defined above. For example, a provider might support "merge" methods declared as follows:
```

Comment:
We had agreement to add lifecycle methods, but not to include the annotations on them. I would like to ensure these annotations are not used up in this manner where they do not seem to add much additional benefit and will not later be available for the spec to use for other purposes.  Some possible revisions to this section follow:

```suggestion
A _lifecycle method_ is an abstract method that accepts an entity as the method's only parameter.
Lifecycle methods allow the program to make changes to persistent data in the data store.

A lifecycle method must:

- have a single parameter whose type is a concrete entity class `E`, and
- be declared `void`, or return exactly the same entity type `E`.

That is, its signature must be:

[source,java]
----
void lifecycle(Entity e);
----

or:

[source,java]
----
Entity lifecycle(Entity e);
----

where `lifecycle` is a method name that starts with `insert`, `update`, `save`, or `delete`, and `Entity` is a concrete entity class.

For example:

[source,java]
----
void insertBook(Book book);
----

[NOTE]
Jakarta Data providers must support lifecycle methods to the extent that the data store is capable of the corresponding operation. If the data store is not capable of the operation, the Jakarta Data provider must raise `UnsupportedOperationException` if the operation is attemped.

There is no special programming model for lifecycle annotations.
The Jakarta Data implementation automatically recognizes the lifecycle annotations it supports.

[NOTE]
====
A Jakarta Data provider might extend this specification to define lifecycle annotations, or to support lifecycle methods with signatures other than the usual signatures defined above. For example, a provider might support "merge" methods declared as follows:
```

We also need to specify the behavior that users can expect from each lifecycle method that is defined from the specification.  There's a pull where I have included this information, but it's been held up over other arguing and hasn't gone in.  After this goes in, that pull will likely be a massive merge conflict and we'll need to cancel it, but maybe we can reuse some of the specifics from there in defining what insert and update are and what the user can expect from them.

Reply:
@njr-11 the last comment I saw from @otaviojava was [this one](https://github.com/jakartaee/data/issues/223#issuecomment-1707757922), where he said:

> I would include the annotations for semantics and flexibility reasons.
>
> Thus, Insert, Update, Delete, and Save annotations.

And I agree on this: the annotations mean that there different kinds of lifecycle methods e.g. `@Save` vs `@Persist` can be technology-specific and not hardcoded in as primitives.

Note that anyone extending `BasicRepository` would not need to use the annotations, since they would just be overriding methods already annotated. (Though we should clarify that, it's not well-defined in what I've written!)

Anyway, let's discuss this more.

Reply:
I don't have any issue with what the annotations would provide, I just don't want them used up for this, which I don't see providing all that much value, because I think we could do different, more valuable things with some of the same annotations later.  In my suggestion, I tried to illustrate how you can achieve basically the same pattern (relying on method name prefix) without the annotations.  This would give us the flexibility to decide on what the annotations will be used for in the future, while remaining compatible with using them for this purpose if we decide to do so.  Vendors could also supply their own annotations in the mean time if they believe it's important.

Reply:
Resolving this per discussion that occurred on this week's Jakarta Data meeting.  I'll re-review it in terms of the decision that was made to use the lifecycle annotations.